fish-rust/src/fd_monitor.rs

```diff
@@ -0,0 +1,573 @@
+        let timeout = self.timeout?;
+        assert!(now >= &last_time, "Steady clock went backwards or bug!");
+        let since = *now - last_time;
+        Some(if since >= timeout {
+            Duration::ZERO
+        } else {
+            timeout - since
+        })
+    }
+
+    /// Invoke this item's callback if its value (when its value is set in the fd or has timed out).
+    /// Returns `true` if the item should be retained or `false` if it should be removed from the
+    /// set.
+    fn service_item(&mut self, fds: &FdReadableSet, now: &Instant) -> ItemAction {
+        let mut result = ItemAction::Retain;
+        let readable = fds.test(self.fd.as_raw_fd());
+        let timed_out = !readable && self.remaining_time(now) == Some(Duration::ZERO);
+        if readable || timed_out {
+            self.last_time = Some(*now);
+            let reason = if readable {
+                ItemWakeReason::Readable
+            } else {
+                ItemWakeReason::Timeout
+            };
+            match &self.callback {
+                FdMonitorCallback::None => panic!("Callback not assigned!"),
+                FdMonitorCallback::Native(callback) => (callback)(&mut self.fd, reason),
+                FdMonitorCallback::Ffi(callback, param) => {
+                    // Safety: identical objects are generated on both sides by cxx bridge as
+                    // integers of the same size (minimum size to fit the enum).
+                    let reason = unsafe { std::mem::transmute(reason) };
+                    (callback)(&mut self.fd as *mut _, reason, *param)
+                }
+            }
+            if !self.fd.is_valid() {
+                result = ItemAction::Remove;
+            }
+        }
+        return result;
+    }
+
+    /// Invoke this item's callback with a poke, if its id is present in the sorted poke list.
+    fn maybe_poke_item(&mut self, pokelist: &[FdMonitorItemId]) -> ItemAction {
+        if self.item_id.0 == 0 || pokelist.binary_search(&self.item_id).is_err() {
+            // Not pokeable or not in the poke list.
+            return ItemAction::Retain;
+        }
+
+        match &self.callback {
+            FdMonitorCallback::None => panic!("Callback not assigned!"),
+            FdMonitorCallback::Native(callback) => (callback)(&mut self.fd, ItemWakeReason::Poke),
+            FdMonitorCallback::Ffi(callback, param) => {
+                // Safety: identical objects are generated on both sides by cxx bridge as
+                // integers of the same size (minimum size to fit the enum).
+                let reason = unsafe { std::mem::transmute(ItemWakeReason::Poke) };
+                (callback)(&mut self.fd as *mut _, reason, *param)
+            }
+        }
+        // Return `ItemAction::Remove` if the callback closed the fd
+        match self.fd.is_valid() {
+            true => ItemAction::Retain,
+            false => ItemAction::Remove,
+        }
+    }
+
+    fn new() -> Self {
```

Comment:
If we want to match the C++ code this should be public.

Reply:
I think this is the kind of thing that doesn’t have to match. that Said it’ll have to be public when we no longer use it via ffi.